File: ts/indicators/core/descriptor.ts
import { Adaptivity } from './adaptivity.js';
import { IndicatorIdentifier } from './indicator-identifier.js';
import { InputRequirement } from './input-requirement.js';
import { OutputDescriptor } from './output-descriptor.js';
import { VolumeUsage } from './volume-usage.js';

/**
 * Classifies an indicator along multiple taxonomic dimensions to enable
 * filtering and display in charting catalogs.
 */
export interface Descriptor {
  /** Uniquely identifies the indicator. */
  identifier: IndicatorIdentifier;

  /** Groups related indicators (e.g., by author or category). */
  family: string;

  /** Whether the indicator adapts its parameters. */
  adaptivity: Adaptivity;

  /** The minimum input data type this indicator consumes. */
  inputRequirement: InputRequirement;

  /** How this indicator uses volume information. */
  volumeUsage: VolumeUsage;

  /** Per-output classifications. */
  outputs: OutputDescriptor[];
}

File: go/indicators/core/descriptor.go
package core

// Descriptor classifies an indicator along multiple taxonomic dimensions to enable
// filtering and display in charting catalogs.
type Descriptor struct {
	// Identifier uniquely identifies the indicator.
	Identifier Identifier `json:"identifier"`

	// Family groups related indicators (e.g., by author or category).
	Family string `json:"family"`

	// Adaptivity classifies whether the indicator adapts its parameters.
	Adaptivity Adaptivity `json:"adaptivity"`

	// InputRequirement is the minimum input data type this indicator consumes.
	InputRequirement InputRequirement `json:"inputRequirement"`

	// VolumeUsage classifies how this indicator uses volume information.
	VolumeUsage VolumeUsage `json:"volumeUsage"`

	// Outputs classifies each output of this indicator.
	Outputs []OutputDescriptor `json:"outputs"`
}

// DescriptorOf returns the taxonomic descriptor for the given indicator identifier.
// The second return value is false if no descriptor is registered for the identifier.
func DescriptorOf(id Identifier) (Descriptor, bool) {
	d, ok := descriptors[id]

	return d, ok
}

// Descriptors returns a copy of the full descriptor registry.
func Descriptors() map[Identifier]Descriptor {
	out := make(map[Identifier]Descriptor, len(descriptors))
	for k, v := range descriptors {
		out[k] = v
	}

	return out
}
